### DISCUSSION TITLE
Updates to speed.python.org UI

### DISCUSSION THREAD

**Discussion by gvanrossum** [faster-cpython/ideas#251] [General]
Everyone who first sees [speed.python.org](https://speed.python.org) is horrified about the UI. It seems typically when people show it to management they end up having to take 20 minutes to an hour to explain how to use the site, what the various graphs and options mean, how to configure it to get the most interesting results. (I know, because I was the manager in this scenario some 6-9 months ago :-).

In this discussion item I will collect UI issues; we will then prioritize and turn them into action items and assign a volunteer to work on the highest priority items.

**Comment by gvanrossum** [comment]
A few things that bothered/confused me:

### Landing page

- [ ] The landing page looks uninviting -- it shows two tables with 40-character base64 strings, rather than a simple graph that you can customize.
- [ ] It is not even clear that the three images at the top of the home page (Changes, Timeline, Comparison) are buttons unless you hover. They look like illustrations.

### Changes page

- [ ] The Changes page could use a brief explanation of what it shows (in fact all pages could use some text explaining what there is to see and do).
- [ ] The Changes page has a CSS bug whereby if the window isn't wide enough the text in the right sidebar overlaps with the main division of the page rather than somehow folding (or causing a horizontal scroll bar to appear).
- [ ] What would really be useful is a table of all the commits "between" two points on the timeline, with links to GitHub, so we can try to figure out a likely cause when we see a surprising jump (up or down) at a given date.

### Timeline page

- [ ] The Timeline page is relatively okay, but should display *something* by default (the default seems to be "None").
- [ ] The Timeline page should have a way to display the "geometric mean" of all benchmarks. This is being output by the pyperformance command in comparison mode, and it would be nice to have this plotted here too (since often this give the best "at a glance" view). It could even be the default graph for the Timeline page.
- [ ] Rename "master" to "main".
- [ ] There is a dropdown to select a second executable (to be drawn in a separate color, apparently) but this is inactive.

### Comparison page

- [ ] Chart colors are arbitrary. It would be nice if each release (3.8, 3.9, 3.10, main) had a color assigned to it so we can recognize it no matter which chart we're looking at, and track it when we're switching branches.
- [ ] It's complicated to select the proper branch and machine, because there are two sets of buttons (one for the machine, one for the branch) but many combinations don't exist. Maybe there should be one big list of all existing combinations.
- [ ] Some benchmarks fail (e.g. greenlet- and genshi-based benchmarks in 3.11 until recently). Those show as zeros, which is confusing -- they should be explicitly marked as "failing" (ideally with a link to a traceback or log).
- [ ] The legend overlaps with the data.
- [ ] The defaults should be different: Horizontal, Normalized (to the newest branch selected, probably?), only main vs. 3.10 branch showing, only one of the three machines (preferably speed.python.org).
- [ ] There should be longer explanations (hover? footnotes?) of what the various hosts mean, and also what some of the branches mean (the "historical" branch is most in need of explanation).
- [ ] There's a non-functional mention of PyPy in the left side bar.
- [ ] Branches should be ordered chronologically, like 3.8, 3.9, 3.10, main.
- [ ] Would be nice if the numbers are also shown numerically.
- [ ] Histogram lines apparently are sensitive to hovering (they change color) but no extra data appears.
- [ ] Fix a bug where if an older branch has a commit newer than the latest commit on main (that it knows of?), it doesn't show the comparison for the older commit.

**Comment by brandtbucher** [comment]
The "All" buttons under the "Executables" and "Benchmarks" sections of the Comparison page sidebar don't work.

**Comment by gvanrossum** [comment]
Sort benchmarks in Comparisons by speedup.
